### PR TITLE
Profile support for dockerEnv() extension 

### PR DESCRIPTION
**Pull request by gintautassulskus**
https://github.com/GoogleContainerTools/minikube-build-tools-for-java/issues/141

Added profile support for dockerEnv() extension

dockerEnv(), dockerEnv("") - refer to default minikube profile
dockerEnv(profile) - refer to custom minikube profile

### CODE REVIEW

File: minikube-gradle-plugin/src/main/java/com/google/cloud/tools/minikube/MinikubeExtension.java [hunk 1]
```diff
@@ -24,9 +24,11 @@
 import java.util.Map;
 import org.gradle.api.Project;
 import org.gradle.api.provider.PropertyState;
+import org.gradle.internal.impldep.com.google.common.base.Preconditions;
```
Comment (chanseokoh) [inline minikube-gradle-plugin/src/main/java/com/google/cloud/tools/minikube/MinikubeExtension.java]: Guava `Preconditions`.

File: minikube-gradle-plugin/src/main/java/com/google/cloud/tools/minikube/MinikubeExtension.java [hunk 2]
```diff
@@ -53,10 +55,26 @@ public void setMinikube(String minikube) {
...
    */
   public Map<String, String> getDockerEnv() throws IOException, InterruptedException {
+    return getDockerEnv("");
+  }
+
+  /**
+   * Gets the minikube docker environment variables by running the command 'minikube docker-env
+   * --shell=none'.
+   *
+   * @param profile Target minikube profile
```
Comment (chanseokoh) [inline minikube-gradle-plugin/src/main/java/com/google/cloud/tools/minikube/MinikubeExtension.java]: nit: "target minikube" (no capitalization)

File: minikube-gradle-plugin/src/main/java/com/google/cloud/tools/minikube/MinikubeExtension.java [hunk 3]
```diff
@@ -53,10 +55,26 @@ public void setMinikube(String minikube) {
...
+   * @return A map of docker environment variables and their values
+   */
+  public Map<String, String> getDockerEnv(String profile) throws IOException, InterruptedException {
+    Preconditions.checkNotNull(profile, "Minikube profile must not be null");
+
     List<String> minikubeDockerEnvCommand =
-        Arrays.asList(minikube.get(), "docker-env", "--shell=none");
+        Arrays.asList(minikube.get(), "docker-env", "--shell=none", "--profile=".concat(profile));
```
Comment (chanseokoh) [inline minikube-gradle-plugin/src/main/java/com/google/cloud/tools/minikube/MinikubeExtension.java]: nit: I think just `"--profile=" + profile`gives us better readbility. Same for the other places.

File: minikube-gradle-plugin/src/test/java/com/google/cloud/tools/minikube/MinikubeExtensionTest.java [hunk 4]
```diff
@@ -16,49 +16,98 @@
...
 /** Tests for MinikubeExtension */
 public class MinikubeExtensionTest {
 
-  @Test
-  public void testGetDockerEnv() throws IOException, InterruptedException {
+  private CommandExecutor commandExecutorMock;
+  private CommandExecutorFactory commandExecutorFactoryMock;
+  private MinikubeExtension minikube;
+
+  List<String> expectedCommand;
```
Comment (chanseokoh) [inline minikube-gradle-plugin/src/test/java/com/google/cloud/tools/minikube/MinikubeExtensionTest.java]: nit: `private`

File: minikube-gradle-plugin/src/test/java/com/google/cloud/tools/minikube/MinikubeExtensionTest.java [hunk 5]
```diff
@@ -16,49 +16,98 @@
...
+    Assert.assertEquals(minikube.getDockerEnv(), minikube.getDockerEnv(profile));
+    verify(commandExecutorMock, times(2)).run(expectedCommand);
+  }
+
+  /*
+   * getDockerEnv() should not permit null values
+   */
+  @Test(expected = NullPointerException.class)
+  public void testGetDockerEnvWithNullProfile() throws IOException, InterruptedException {
+    minikube.getDockerEnv(null);
```
Comment (chanseokoh) [inline minikube-gradle-plugin/src/test/java/com/google/cloud/tools/minikube/MinikubeExtensionTest.java]: Remove `(expected = ...)` and turn this into

```java
try {
  minikube.getDockerEnv(null);
  Assert.fail("some message if not null");
} catch (NullPointerException ex) {
  Assert.assertNotNull(ex.getMessage());
  // You can actually make Preconditions return some useful message and
  // do Assert.assertEquals("message", ex.getMessage());
}

File: minikube-gradle-plugin/src/test/java/com/google/cloud/tools/minikube/MinikubeExtensionTest.java [hunk 6]
```diff
@@ -16,49 +16,98 @@
...
+
+  static {
+    Map<String, String> map = new HashMap<>(2);
+    map.put("ENV_VAR1", "VAL1");
+    map.put("ENV_VAR2", "VAL2");
+    expectedMap = Collections.unmodifiableMap(map);
+  }
+
+  /*
+   * Reinitialise variables before each test.
```
Comment (chanseokoh) [inline minikube-gradle-plugin/src/test/java/com/google/cloud/tools/minikube/MinikubeExtensionTest.java]: nit: I think it's unnecessary to add this comment blocks in these test methods.

File: minikube-gradle-plugin/src/test/java/com/google/cloud/tools/minikube/MinikubeExtensionTest.java [hunk 7]
```diff
@@ -16,49 +16,98 @@
...
 
-    // Defined the expected command to run, its output, and the resulting docker-env map.
-    List<String> expectedCommand =
-        Arrays.asList("/test/path/to/minikube", "docker-env", "--shell=none");
-    List<String> dockerEnvOutput = Arrays.asList("ENV_VAR1=VAL1", "ENV_VAR2=VAL2");
-    Map<String, String> expectedMap = new HashMap<>(2);
-    expectedMap.put("ENV_VAR1", "VAL1");
-    expectedMap.put("ENV_VAR2", "VAL2");
+    expectedCommand =
+        new LinkedList<>(Arrays.asList("/test/path/to/minikube", "docker-env", "--shell=none"));
```
Comment (chanseokoh) [inline minikube-gradle-plugin/src/test/java/com/google/cloud/tools/minikube/MinikubeExtensionTest.java]: nit: probably `ArrayList` will just work.

**Comment by gintautassulskus** [comment]
Thanks for the useful remarks! Addressed.

File: minikube-gradle-plugin/README.md [hunk 8]
```diff
@@ -41,13 +41,14 @@ task minikubeCustom(type: com.google.cloud.tools.minikube.MinikubeTask) {
...
-(See [`docker-machine env` reference](https://docs.docker.com/machine/reference/env/))
-
+The `minikube` extension also provides a method to obtain the environment variables necessary for working with minikube's docker daemon (See [`docker-machine env` reference](https://docs.docker.com/machine/reference/env/))
+:
 ```
-def minikubeDockerEnv = minikube.getDockerEnv()
+def minikubeDockerEnv = minikube.getDockerEnv(<profile>)
 ```
 
+Optional parameter `<profile>` sets target minikube instance (minikube --profile <proile>).
```
Comment (loosebazooka) [inline minikube-gradle-plugin/README.md]: Spelling: `<proile>` -> `<profile>`

**Comment by gintautassulskus** [comment]
@loosebazooka, I could not find any documentation on this. Here's the issue for documentation request: https://github.com/kubernetes/minikube/issues/2371

Could just link to that issue.

**Comment by loosebazooka** [comment]
@gintautassulskus I'm a little curious about what the exact behavior is, I'll try talking to the minikube team and see if I can get a followup on that issue.

**Comment by gintautassulskus** [comment]
If you have a particular question,  let me know. I may be able to answer it as well. Our development workflow relies on profile feature but unfortunately not all the tools support it. So far `bash it all!` solution worked out best.

File: minikube-gradle-plugin/build.gradle [hunk 9]
```diff
@@ -16,6 +16,7 @@ repositories {
 
 dependencies {
   compile gradleApi()
+  compile 'com.google.guava:guava:23.5-jre'
```
Comment (loosebazooka) [inline minikube-gradle-plugin/build.gradle]: I think I'd rather not add this dependency if we can avoid it. We're trying to keep dependencies to a minimum as buildscript classpath level dependencies tend to have weird behavior when they clash in gradle. Looks like we can just replace the preconditions check in MinikubeExtension with simple custom code that throws NullPointerException?

File: minikube-gradle-plugin/src/test/java/com/google/cloud/tools/minikube/MinikubeExtensionTest.java [hunk 10]
```diff
@@ -17,48 +17,103 @@
...
 import static org.mockito.Mockito.when;
 
 import com.google.cloud.tools.minikube.util.CommandExecutor;
 import com.google.cloud.tools.minikube.util.CommandExecutorFactory;
 import java.io.IOException;
-import java.util.Arrays;
-import java.util.HashMap;
-import java.util.List;
-import java.util.Map;
+import java.util.*;
```
Comment (loosebazooka) [inline minikube-gradle-plugin/src/test/java/com/google/cloud/tools/minikube/MinikubeExtensionTest.java]: Ah, this part needs to be expanded back out too.

Comment (gintautassulskus) [reply minikube-gradle-plugin/src/test/java/com/google/cloud/tools/minikube/MinikubeExtensionTest.java]: Done.
Thanks for the thorough review. Was quite a good induction to Google's code style practices; I was not aware of the tools.

**Pull request merged**